File: q0003/q0003.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int longestLength = 0;
        unordered_map<char, int> charIndex;
        int l = 0;
        for (int r = 0; r < s.size(); r++)
        {
            char currChar = s[r];
            if (charIndex.find(currChar) != charIndex.end())
            {
                l = max(l, charIndex[currChar] + 1);
            }
            longestLength = max(longestLength, r - l + 1);
            charIndex[currChar] = r;
        }
        return longestLength;
    }
};

File: q0003/q0003.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longestLength = 0;
        int l = 0;
        HashMap<Character, Integer> charIndex = new HashMap<>();
        for (int r = 0; r < s.length(); r++) {
            char currChar = s.charAt(r);
            if (charIndex.containsKey(currChar)) {
                l = Math.max(l, charIndex.get(currChar) + 1);
            }
            longestLength = Math.max(longestLength, r - l + 1);
            charIndex.put(currChar, r);
        }
        return longestLength;
    }
}